src/NUnitFramework/framework/Internal/Builders/CombinatorialStrategy.cs

```diff
@@ -40,7 +40,16 @@ public class CombinatorialStrategy : ICombiningStrategy
         /// <returns>The test cases.</returns>
         public IEnumerable<ITestCaseData> GetTestCases(IEnumerable[] sources)
         {
-            List<ITestCaseData> testCases = new List<ITestCaseData>();
+            return GetCombinations<ITestCaseData>(sources, (o) => new TestCaseParameters(o));
+        }
+
+        /// <summary>
+        /// Returns an <see cref="IEnumerable{T}"/> representing every single possible combination of the given parameters
+        /// </summary>
+        /// <returns>Every combination of parameters</returns>
```

Comment:
Leaving this method internal is an option I'd consider.
If you keep it public, I'd like to consistently use full sentences (including end marks) for XML documentation. Also, we recommend not adding `<returns>` since it isn't shown in intellisense and our XML docs aren't reused for anything else.

Reply:
The argument for leaving it public would be that, in cases where TestCaseData is extended upon in a derived class, you will not be able to use TestCaseData.Combinatorial() for that derived class since it would be returning TestCaseData instances and not the instance of the derived class.  So, in order to get Combinatorial() behavior in a derived TestCaseData instance, you would have to add your own static method that calls CombinatorialStrategy.GetCombinations().

I guess an alternative to solving this would be to expose an overload for TestCaseData.Combinatorial() as a protected static method where it would be something like this:

```csharp
public class TestCaseData : TestCaseParameters
{
...
     protected static IEnumerable<T> Combinatorial(IEnumerable[] values, Func<object[], T> handle)
     {
          return CombinatorialStrategy.GetCombinations<T>(values, handle);
     }
...
}
```
And then in the derived class, you can use the ```new``` keyword to hide away the old Combinatorial method and expose the new one, like so:
```csharp
public class DerivedTestCaseData : TestCaseData
{
...
     public static new IEnumerable<DerivedTestCaseData> Combinatorial(IEnumerable[] values)
     {
          return TestCaseData.Combinatorial<DerivedTestCaseData>(values, (o) => new DerivedTestCaseData(o));
     }
...
}
```

If you'd prefer, i can go ahead and make those changes, please let me know though.

Reply:
I agree that there should be some unit tests.  I ran into some issues even trying to run existing unit tests in the solution, though.

I can try again once i get a response on the other items of note here.

Reply:
I like it public then. That makes sense. If it's a public API though, we don't want it in the Internal namespace. To me, it also doesn't seem very discoverable where it is.

When we decide how to expose this API, I'd like to do it in a consistent way that achieves the goal @CharliePoole mentioned:

> IMO it would be nice if all combining strategies could be made available, not just Combinatorial.

If we make it public in this pull request, it will probably take longer to merge since @nunit/framework-team needs to decide what the full picture will look like. Are you okay with that?